src/main/java/slimeknights/tconstruct/plugin/jei/ToolBuildingCategory.java

```diff
@@ -0,0 +1,147 @@
+    this.anvil = guiHelper.createDrawable(BACKGROUND_LOC, 128, 61, ITEM_SIZE, ITEM_SIZE);
+  }
+
+  @Override
+  public void setRecipe(IRecipeLayoutBuilder builder, ToolBuildingRecipe recipe, IFocusGroup focuses) {
+    List<IToolPart> parts = recipe.getToolParts();
+    List<ItemStack> extraItems = recipe.getIngredients().stream().flatMap(ingredient -> Arrays.stream(ingredient.getItems())).toList();
+    List<LayoutSlot> layoutSlots = recipe.getLayoutSlots();
+
+    // every part and its material variants
+    List<ItemStack> allInputParts = parts.stream()
+        .flatMap(part -> MaterialRegistry.getInstance().getVisibleMaterials().stream()
+          .filter(part::canUseMaterial)
+          .flatMap(mat -> Stream.of(part.withMaterial(mat.getIdentifier()))))
+        .toList();
+
+    if (parts.size() + extraItems.size() != layoutSlots.size()) {
+      TConstruct.LOG.error(String.format("Tool part count and layout slot count for %s do not match!", recipe.getId()));
+      return;
+    }
+
+    builder.addInvisibleIngredients(RecipeIngredientRole.INPUT).addItemStacks(allInputParts);
+    builder.addInvisibleIngredients(RecipeIngredientRole.INPUT).addItemStacks(extraItems);
+
+    Vec2 offsets = getOffsets(layoutSlots);
+    for (int i = 0; i < parts.size(); i++) {
+      builder.addSlot(RecipeIngredientRole.INPUT, (int) (layoutSlots.get(i).getX() + offsets.x), (int) (layoutSlots.get(i).getY() + offsets.y))
+             .addItemStack(getDisplayPart(parts.get(i), i));
+    }
+
+    if (recipe.getOutput() instanceof IModifiableDisplay modifiable) {
+      builder.addSlot(RecipeIngredientRole.OUTPUT, WIDTH - 23, 23)
+        .addItemStack(modifiable.getRenderTool());
+    }
+  }
+
+  @Override
+  public void draw(ToolBuildingRecipe recipe, IRecipeSlotsView recipeSlotsView, PoseStack stack, double mouseX, double mouseY) {
+    if (recipe.isBroadTool()) {
+      this.anvil.draw(stack, 73, 42);
+    }
+
+    List<LayoutSlot> layoutSlots = recipe.getLayoutSlots();
+    Vec2 offsets = getOffsets(layoutSlots);
+
+    for (LayoutSlot slot : layoutSlots) {
+      // need to offset by 1 because the inventory slot icons are 18x18
+      this.slot.draw(stack, (int) (slot.getX() + offsets.x - 1), (int) (slot.getY() + offsets.y - 1));
+    }
+  }
+
+  @Override
+  public List<Component> getTooltipStrings(ToolBuildingRecipe recipe, IRecipeSlotsView recipeSlotsView, double mouseX, double mouseY) {
+    return recipe.isBroadTool() && GuiUtil.isHovered((int) mouseX, (int) mouseY, 73, 42, ITEM_SIZE, ITEM_SIZE) ?
+      List.of(TConstruct.makeTranslation("jei", "tinkering.tool_building.anvil")) :
+      List.of();
+  }
+
+  // use display parts to be more consistent
+  private ItemStack getDisplayPart(IToolPart toolPart, int i) {
+      ItemStack item = toolPart.withMaterialForDisplay(ToolBuildHandler.getRenderMaterial(i));
+      item.getOrCreateTag().putBoolean(TooltipUtil.KEY_DISPLAY, true);
+      return item;
+  }
```

Comment:
Is this not a helper method already? Pretty sure we do this in the books. If its not a helper method already, make it one in `ToolBuildHandler` so this code is only in one place (I believe its `ContentTool` or `ToolContent` (whatever that class is called) that would be doing this already)

Reply:
It is used in `ContentTool` (where I borrowed it from), but its not a separate method.